### ISSUE TITLE
type_repetition_in_bounds false positive with associated types

### ISSUE THREAD

**Issue by ecton** [rust-lang/rust-clippy#8590]
### Summary

When using a trait with an associated type alongside a generic with the same name, attempting to assign bounds to both causes a false positive of the `type_repition_in_bounds`.

### Lint Name

type_repetition_in_bounds

### Reproducer

I tried this code:

```rust
#![deny(clippy::type_repetition_in_bounds)]

use std::marker::PhantomData;

trait SomeTrait<B> {
    type Type;
}

struct Foo<ST, Type>(PhantomData<(ST, Type)>);

impl<ST, Type> From<()> for Foo<ST, Type>
where
    ST: SomeTrait<Type>,
    ST::Type: Sized,
    Type: Sized,
{
    fn from(_: ()) -> Self {
        Self(PhantomData)
    }
}
```

I saw this happen:

```
error: this type has already been used as a bound predicate
  --> src/lib.rs:15:5
   |
15 |     Type: Sized,
   |     ^^^^^^^^^^^
   |
note: the lint level is defined here
  --> src/lib.rs:1:9
   |
1  | #![deny(clippy::type_repetition_in_bounds)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: consider combining the bounds: `Type: Sized + Sized`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#type_repetition_in_bounds
```

I expected to see this happen:

No clippy warnings.

### Version

```text
rustc 1.59.0 (9d1b2106e 2022-02-23)
binary: rustc
commit-hash: 9d1b2106e23b1abd32fce1f17267604a5102f57a
commit-date: 2022-02-23
host: x86_64-unknown-linux-gnu
release: 1.59.0
LLVM version: 13.0.0
```
